### PR TITLE
Add --pbsFlags feature

### LINKED ISSUE #109: Add the option to add custom PBS flags.
Add `--pbs-flags` to the command line to allow custom PBS flags.

### PR DESCRIPTION
**Pull request by mgermain**
Add the --pbsFlags feature and some tests.
Fix #109

### CODE REVIEW

File: scripts/smart-dispatch [hunk 1]
```diff
@@ -116,6 +116,8 @@ def main():
                       }
 
     job_generator = job_generator_factory(queue, commands, command_params, CLUSTER_NAME, path_job)
+    if args.pbsFlags:
```
Comment (MarcCote) [inline scripts/smart-dispatch]: Should be `if args.pbsFlags is not None`. (see http://stackoverflow.com/questions/7816363/if-a-vs-if-a-is-not-none)

File: scripts/smart-dispatch [hunk 2]
```diff
@@ -149,9 +151,10 @@ def parse_arguments():
     parser.add_argument('-f', '--commandsFile', type=file, required=False, help='File containing commands to launch. Each command must be on a seperate line. (Replaces commandAndOptions)')
 
     parser.add_argument('-l', '--modules', type=str, required=False, help='List of additional modules to load.', nargs='+')
-    parser.add_argument('-x', '--doNotLaunch', action='store_true', help='Creates the QSUB files without launching them.')
+    parser.add_argument('-x', '--doNotLaunch', action='store_true', help='Generate all the files without launching the job.')
 
     parser.add_argument('-p', '--pool', type=int, help="Number of workers that will be consuming commands. Default: Nb commands")
+    parser.add_argument('--pbsFlags', type=str, help='ADVANCED USAGE: Allow to pass a space seperates list of PBS flags. Ex:--pbsFlags="-lfeature=k80 -t0-4"')
```
Comment (MarcCote) [inline scripts/smart-dispatch]: separates -> separated

File: smartdispatch/job_generator.py [hunk 3]
```diff
@@ -47,9 +47,23 @@ def __init__(self, queue, commands, command_params={}, base_path="./"):
             self._add_cluster_specific_rules()
 
     def add_pbs_flags(self, flags):
+        ressources = {}
```
Comment (MarcCote) [inline smartdispatch/job_generator.py]: re**s**sources -> resources

File: smartdispatch/job_generator.py [hunk 4]
```diff
@@ -47,9 +47,23 @@ def __init__(self, queue, commands, command_params={}, base_path="./"):
...
+        options = {}
+
         for flag in flags:
+            flag = flag
             if flag.startswith('-l'):
-                flag[2:].trim()
+                ressource = flag[2:]
```
Comment (MarcCote) [inline smartdispatch/job_generator.py]: re**s**source -> resource

File: smartdispatch/job_generator.py [hunk 5]
```diff
@@ -42,7 +42,30 @@ def __init__(self, queue, commands, command_params={}, base_path="./"):
         self.nb_gpus_per_command = command_params.get('nb_gpus_per_command', 1)
         #self.mem_per_command = command_params.get('mem_per_command', 0.0)
 
-    def generate_pbs(self):
+        self.pbs_list = self._generate_base_pbs()
+        if type(self) is not JobGenerator:
+            self._add_cluster_specific_rules()
+
+    def add_pbs_flags(self, flags):
+        ressources = {}
```
Comment (MarcCote) [inline smartdispatch/job_generator.py]: re**s**sources -> resources

File: smartdispatch/job_generator.py [hunk 6]
```diff
@@ -42,7 +42,30 @@ def __init__(self, queue, commands, command_params={}, base_path="./"):
...
+        options = {}
+
+        for flag in flags:
+            flag = flag
+            if flag.startswith('-l'):
+                ressource = flag[2:]
```
Comment (MarcCote) [inline smartdispatch/job_generator.py]: re**s**source -> resource

File: smartdispatch/tests/test_job_generator.py [hunk 7]
```diff
@@ -89,6 +91,40 @@ def test_write_pbs_files(self):
         filenames = job_generator.write_pbs_files(self.testing_dir)
         assert_equal(len(filenames), 4)
 
+    def _test_add_pbs_flags(self, flags):
+        job_generator = JobGenerator(self.queue, self.commands)
+        job_generator.add_pbs_flags(flags)
+
+        ressources = []
```
Comment (MarcCote) [inline smartdispatch/tests/test_job_generator.py]: re**s**sources -> resources

File: smartdispatch/tests/test_job_generator.py [hunk 8]
```diff
@@ -95,28 +90,66 @@ def test_write_pbs_files(self):
         filenames = job_generator.write_pbs_files(self.testing_dir)
         assert_equal(len(filenames), 4)
 
+    def _test_add_pbs_flags(self, flags):
+        job_generator = JobGenerator(self.queue, self.commands)
+        job_generator.add_pbs_flags(flags)
 
-class TestGuilliminQueue(TestJobGenerator):
+        ressources = []
```
Comment (MarcCote) [inline smartdispatch/tests/test_job_generator.py]: re**s**sources -> resources

File: smartdispatch/tests/test_job_generator.py [hunk 9]
```diff
@@ -95,28 +90,66 @@ def test_write_pbs_files(self):
...
+            pbs_str = pbs.__str__()
+            for flag in ressources:
+                assert_equal(pbs_str.count(flag), 1)
+                assert_equal(pbs_str.count(flag[:flag.find('=')]), 1)
+            for flag in options:
+                assert_equal(pbs_str.count(flag), 1)
+
+    def test_add_pbs_flags(self):
+        for flag in self.pbs_flags:
+            yield self._test_add_pbs_flags, [flag]
```
Comment (MarcCote) [inline smartdispatch/tests/test_job_generator.py]: Can you really yield in a test?

Comment (mgermain) [inline smartdispatch/tests/test_job_generator.py]: Yes. That's the proper way with nose to generate parameterized tests. In other words, write a test only once and have it run on different data. Look at the output it's quite nice.

File: scripts/smart-dispatch (hunk 2 above)
Comment (mgermain) [inline scripts/smart-dispatch]: So many typos!

File: smartdispatch/tests/test_job_generator.py (hunk 9 above)
Comment (MarcCote) [inline smartdispatch/tests/test_job_generator.py]: Nice.

**Comment by MarcCote** [comment]
Good, now we can "hack" our PBS files like experts ;)

**Pull request merged**